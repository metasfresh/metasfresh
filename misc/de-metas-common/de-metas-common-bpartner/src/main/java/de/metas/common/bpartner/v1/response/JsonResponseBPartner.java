/*
 * #%L
 * de-metas-common-bpartner
 * %%
 * Copyright (C) 2021 metas GmbH
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program. If not, see
 * <http://www.gnu.org/licenses/gpl-2.0.html>.
 * #L%
 */

package de.metas.common.bpartner.v1.response;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonInclude.Include;
import com.fasterxml.jackson.annotation.JsonProperty;
import de.metas.common.changelog.JsonChangeInfo;
import de.metas.common.rest_api.common.JsonMetasfreshId;
import de.metas.common.rest_api.common.JsonExternalId;
import io.swagger.annotations.ApiModel;
import io.swagger.annotations.ApiModelProperty;
import lombok.Builder;
import lombok.NonNull;
import lombok.Value;

import javax.annotation.Nullable;

@Value
@ApiModel(description = "Note that given the respective use-case, either one of both properties migh be `null`, but not both at once.")
public class JsonResponseBPartner
{
	public static final String GROUP_NAME = "group";
	public static final String URL = "url";
	public static final String URL_2 = "url2";
	public static final String URL_3 = "url3";
	public static final String LANGUAGE = "language";
	public static final String PHONE = "phone";
	public static final String PARENT_ID = "parentId";
	public static final String COMPANY_NAME = "companyName";
	public static final String NAME = "name";
	public static final String NAME_2 = "name2";
	public static final String NAME_3 = "name3";
	public static final String EXTERNAL_ID = "externalId";
	public static final String METASFRESH_ID = "metasfreshId";
	public static final String CODE = "code";
	public static final String GLOBAL_ID = "globalId";
	public static final String ACTIVE = "active";
	public static final String VENDOR = "vendor";
	public static final String CUSTOMER = "customer";
	public static final String COMPANY = "company";
	public static final String VAT_ID = "vatId";

	private static final String CHANGE_INFO = "changeInfo";

	@ApiModelProperty( //
			required = true, //
			dataType = "java.lang.Integer", //
			value = "This translates to `C_BPartner.C_BPartner_ID`.")
	@JsonProperty(METASFRESH_ID)
	@JsonInclude(Include.NON_NULL)
	JsonMetasfreshId metasfreshId;

	@ApiModelProperty( //
			required = false, //
			dataType = "java.lang.String", //
			value = "This translates to `C_BPartner.ExternalId`.")
	@JsonProperty(EXTERNAL_ID)
	@JsonInclude(Include.NON_NULL)
	JsonExternalId externalId;

	@ApiModelProperty(required = false, value = "This translates to `C_BPartner.Value`.")
	@JsonProperty(CODE)
	String code;

	@ApiModelProperty(required = false, value = "This translates to `C_BPartner.GlobalId`.")
	@JsonProperty(GLOBAL_ID)
	String globalId;

	@ApiModelProperty(required = false, value = "This translates to `C_BPartner.IsActive`.")
	@JsonProperty(ACTIVE)
	boolean active;

	@ApiModelProperty(required = false, value = "This translates to `C_BPartner.Name`.")
	@JsonProperty(NAME)
	String name;

	@ApiModelProperty(required = false, value = "This translates to `C_BPartner.Name2`.")
	@JsonProperty(NAME_2)
	@JsonInclude(Include.NON_NULL)
	String name2;

	@ApiModelProperty(required = false, value = "This translates to `C_BPartner.Name3`.")
	@JsonProperty(NAME_3)
	@JsonInclude(Include.NON_NULL)
	String name3;

	@ApiModelProperty( //
			required = false, //
			value = "This translates to `C_BPartner.CompanyName`.\n"
					+ "If set, the the respective `C_BPartner` record will also have `IsCompany='Y'`")
	@JsonProperty(COMPANY_NAME)
	@JsonInclude(Include.NON_NULL)
	String companyName;

	@ApiModelProperty( //
			required = false, //
			value = "This translates to `C_BPartner.BPartner_Parent_ID`. It's a this bpartner's central/parent company",//
			dataType = "java.lang.Integer")
	@JsonProperty(PARENT_ID)
	@JsonInclude(Include.NON_NULL)
	JsonMetasfreshId parentId;

	@ApiModelProperty( //
			required = false, //
			value = "This translates to `C_BPartner.Phone2`. It's this bpartner's central phone number")
	@JsonProperty(PHONE)
	@JsonInclude(Include.NON_NULL)
	String phone;

	@JsonInclude(Include.NON_NULL)
	@JsonProperty(LANGUAGE)
	String language;

	@ApiModelProperty(required = false, value = "This translates to `C_BPartner.URL`.")
	@JsonProperty(URL)
	@JsonInclude(Include.NON_NULL)
	String url;

	@ApiModelProperty(required = false, value = "This translates to `C_BPartner.URL2`.")
	@JsonProperty(URL_2)
	@JsonInclude(Include.NON_NULL)
	private String url2;

	@ApiModelProperty(required = false, value = "This translates to `C_BPartner.URL3`.")
	@JsonProperty(URL_3)
	@JsonInclude(Include.NON_NULL)
	private String url3;

	@ApiModelProperty( //
			required = false, //
			value = "Name of the business partner's group")
	@JsonProperty(GROUP_NAME)
	@JsonInclude(Include.NON_NULL)
	String group;

	@ApiModelProperty(required = false, value = "This translates to `C_BPartner.IsVendor`.")
	@JsonProperty(VENDOR)
	boolean vendor;

	@ApiModelProperty(required = false, value = "This translates to `C_BPartner.IsCustomer`.")
	@JsonProperty(CUSTOMER)
	boolean customer;

	@ApiModelProperty(required = false, value = "This translates to `C_BPartner.IsCompany`.")
	@JsonProperty(COMPANY)
	boolean company;

	@ApiModelProperty(required = false, value = "This translates to `C_BPartner.VATaxID`.")
	@JsonProperty(VAT_ID)
	String vatId;

	@ApiModelProperty(position = 9999) // shall be last
	@JsonProperty(CHANGE_INFO)
	@JsonInclude(Include.NON_NULL)
	JsonChangeInfo changeInfo;

	@JsonCreator
	@Builder(toBuilder = true)
	private JsonResponseBPartner(
			@JsonProperty(METASFRESH_ID) @NonNull final JsonMetasfreshId metasfreshId,
			@JsonProperty(EXTERNAL_ID) @Nullable final JsonExternalId externalId,
			@JsonProperty(CODE) @Nullable final String code, // usually metasfresh makes sure it's not null; but in unit-tests it might be; also, let's just return the result here, and avoid throwing an NPE
			@JsonProperty(GLOBAL_ID) @Nullable final String globalId,
			@JsonProperty(ACTIVE) @NonNull final Boolean active,
			@JsonProperty(NAME) @NonNull final String name,
			@JsonProperty(NAME_2) @Nullable final String name2,
			@JsonProperty(NAME_3) @Nullable final String name3,
			@JsonProperty(COMPANY_NAME) @Nullable final String companyName,
			@JsonProperty(PARENT_ID) @Nullable final JsonMetasfreshId parentId,
			@JsonProperty(PHONE) @Nullable final String phone,
			@JsonProperty(LANGUAGE) @Nullable final String language,
			@JsonProperty(URL) @Nullable final String url,
			@JsonProperty(URL_2) @Nullable final String url2,
			@JsonProperty(URL_3) @Nullable final String url3,
			@JsonProperty(GROUP_NAME) @Nullable final String group,
			@JsonProperty(VENDOR) @NonNull final Boolean vendor,
			@JsonProperty(CUSTOMER) @NonNull final Boolean customer,
			@JsonProperty(COMPANY) @NonNull final Boolean company,
			@JsonProperty(VAT_ID) @Nullable final String vatId,
			//
			@JsonProperty(CHANGE_INFO) @Nullable JsonChangeInfo changeInfo)
	{
		this.metasfreshId = metasfreshId;
		this.externalId = externalId;
		this.code = code;
		this.globalId = globalId;
		this.active = active;

		this.name = name;
		this.name2 = name2;
		this.name3 = name3;

		this.companyName = companyName;

		this.parentId = parentId;

		this.phone = phone;
		this.language = language;

		this.url = url;
		this.url2 = url2;
		this.url3 = url3;

		this.group = group;

		this.vendor = vendor;
		this.customer = customer;
		this.company = company;

		this.vatId = vatId;

		this.changeInfo = changeInfo;
	}
}
