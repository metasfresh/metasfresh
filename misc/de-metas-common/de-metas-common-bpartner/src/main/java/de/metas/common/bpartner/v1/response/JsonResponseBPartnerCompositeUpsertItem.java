/*
 * #%L
 * de-metas-common-bpartner
 * %%
 * Copyright (C) 2021 metas GmbH
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program. If not, see
 * <http://www.gnu.org/licenses/gpl-2.0.html>.
 * #L%
 */

package de.metas.common.bpartner.v1.response;

import java.util.List;

import javax.annotation.Nullable;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonInclude.Include;

import io.swagger.annotations.ApiModelProperty;

import com.fasterxml.jackson.annotation.JsonProperty;

import lombok.Builder;
import lombok.Singular;
import lombok.Value;

@Value
public class JsonResponseBPartnerCompositeUpsertItem
{
	@ApiModelProperty(position = 10)
	@JsonInclude(Include.NON_NULL)
	JsonResponseUpsertItem responseBPartnerItem;

	@ApiModelProperty(position = 20)
	@JsonInclude(Include.NON_EMPTY)
	List<JsonResponseUpsertItem> responseLocationItems;

	@ApiModelProperty(position = 30)
	@JsonInclude(Include.NON_EMPTY)
	List<JsonResponseUpsertItem> responseContactItems;

	@ApiModelProperty(position = 40)
	@JsonInclude(Include.NON_EMPTY)
	List<JsonResponseUpsertItem> responseBankAccountItems;

	@Builder
	@JsonCreator
	public JsonResponseBPartnerCompositeUpsertItem(
			@JsonProperty("responseBPartnerItem") @Nullable final JsonResponseUpsertItem responseBPartnerItem,
			@JsonProperty("responseLocationItems") @Singular final List<JsonResponseUpsertItem> responseLocationItems,
			@JsonProperty("responseContactItems") @Singular final List<JsonResponseUpsertItem> responseContactItems,
			@JsonProperty("responseBankAccountItems") @Singular final List<JsonResponseUpsertItem> responseBankAccountItems)
	{
		this.responseBPartnerItem = responseBPartnerItem;
		this.responseLocationItems = responseLocationItems;
		this.responseContactItems = responseContactItems;
		this.responseBankAccountItems = responseBankAccountItems;
	}


}
