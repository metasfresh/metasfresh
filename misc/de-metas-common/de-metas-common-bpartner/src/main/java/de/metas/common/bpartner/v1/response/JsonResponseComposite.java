/*
 * #%L
 * de-metas-common-bpartner
 * %%
 * Copyright (C) 2021 metas GmbH
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program. If not, see
 * <http://www.gnu.org/licenses/gpl-2.0.html>.
 * #L%
 */

package de.metas.common.bpartner.v1.response;

import static de.metas.common.util.CoalesceUtil.coalesce;

import java.util.List;
import java.util.Objects;

import com.fasterxml.jackson.annotation.JsonAutoDetect;
import com.fasterxml.jackson.annotation.JsonAutoDetect.Visibility;
import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonInclude.Include;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.google.common.collect.ImmutableList;

import de.metas.common.util.EmptyUtil;
import de.metas.common.rest_api.common.JsonExternalId;
import io.swagger.annotations.ApiModel;
import io.swagger.annotations.ApiModelProperty;
import lombok.Builder;
import lombok.NonNull;
import lombok.Singular;
import lombok.Value;

@ApiModel(description = "A BPartner with `n` contacts and `n` locations.\n" //
		+ "Note that given the respective use-case, either `bpartner.code` `bpartner.externalId` might be `null`, but not both at once.")
@Value
@JsonAutoDetect(fieldVisibility = Visibility.ANY, getterVisibility = Visibility.NONE, isGetterVisibility = Visibility.NONE, setterVisibility = Visibility.NONE)
public class JsonResponseComposite
{
	@ApiModelProperty(required = true, value = "The `AD_Org.Value` of the `C_BPartner`'s AD_Org_ID")
	@JsonInclude(Include.NON_NULL)
	String orgCode;

	JsonResponseBPartner bpartner;

	@ApiModelProperty(value = "The location's GLN can be used to lookup the whole bpartner; if nultiple locations with GLN are provided, then only the first one is used")
	@JsonInclude(Include.NON_EMPTY)
	List<JsonResponseLocation> locations;

	@JsonInclude(Include.NON_EMPTY)
	List<JsonResponseContact> contacts;

	@Builder(toBuilder = true)
	@JsonCreator
	private JsonResponseComposite(
			@JsonProperty("orgCode") @NonNull final String orgCode,
			@JsonProperty("bpartner") @NonNull final JsonResponseBPartner bpartner,
			@JsonProperty("locations") @Singular final List<JsonResponseLocation> locations,
			@JsonProperty("contacts") @Singular final List<JsonResponseContact> contacts)
	{
		this.orgCode = orgCode;
		this.bpartner = bpartner;
		this.locations = coalesce(locations, ImmutableList.of());
		this.contacts = coalesce(contacts, ImmutableList.of());

		final boolean lokupValuesAreOk = !EmptyUtil.isEmpty(bpartner.getCode(), true)
				|| bpartner.getExternalId() != null
				|| !extractLocationGlns().isEmpty();
		if (!lokupValuesAreOk)
		{
			throw new RuntimeException("At least one of bpartner.code, bpartner.externalId or one location.gln needs to be non-empty; this=" + this);
		}
	}

	public ImmutableList<String> extractLocationGlns()
	{
		return this.locations
				.stream()
				.map(JsonResponseLocation::getGln)
				.filter(gln -> !EmptyUtil.isEmpty(gln, true))
				.collect(ImmutableList.toImmutableList());
	}

	public JsonResponseComposite withExternalId(@NonNull final JsonExternalId externalId)
	{
		if (Objects.equals(externalId, bpartner.getExternalId()))
		{
			return this; // nothing to do
		}
		return toBuilder()
				.bpartner(bpartner.toBuilder().externalId(externalId).build())
				.build();
	}
}
