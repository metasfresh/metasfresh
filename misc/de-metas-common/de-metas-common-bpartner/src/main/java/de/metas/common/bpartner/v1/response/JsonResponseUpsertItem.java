/*
 * #%L
 * de-metas-common-bpartner
 * %%
 * Copyright (C) 2021 metas GmbH
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program. If not, see
 * <http://www.gnu.org/licenses/gpl-2.0.html>.
 * #L%
 */

package de.metas.common.bpartner.v1.response;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import de.metas.common.rest_api.common.JsonMetasfreshId;
import io.swagger.v3.oas.annotations.media.Schema;
import lombok.Builder;
import lombok.NonNull;
import lombok.Value;

import javax.annotation.Nullable;

@Schema(description = "Response to a a single mater data upsert request entity")
@Value
public class JsonResponseUpsertItem
{
	@Schema(enumAsRef = true, description = "SyncOutcome: \n" +
			"* `CREATED`\n" +
			"* `UPDATED` - Master data was updated; note that it's possible that nothing really changed due to the update.\n" +
			"* `NOTHING_DONE` - E.g. if a location already exists and the sync advise is to not change existing entities.\n" +
			"")
	public enum SyncOutcome
	{
		CREATED,

		UPDATED,

		NOTHING_DONE
	}

	@Schema(description = "The identifier that was specified in the repective upsert request")
	String identifier;

	@Schema(description = "The metasfresh-ID of the upserted record.\n"
			+ "Can be null if the respective resource did not exist and the sync-advise indicated to do nothing.")
	JsonMetasfreshId metasfreshId;

	SyncOutcome syncOutcome;

	@Builder
	@JsonCreator
	public JsonResponseUpsertItem(
			@JsonProperty("identifier") @NonNull final String identifier,
			@JsonProperty("metasfreshId") @Nullable final JsonMetasfreshId metasfreshId,
			@JsonProperty("syncOutcome") @NonNull final SyncOutcome syncOutcome)
	{
		this.identifier = identifier;
		this.metasfreshId = metasfreshId;
		this.syncOutcome = syncOutcome;
	}

}
