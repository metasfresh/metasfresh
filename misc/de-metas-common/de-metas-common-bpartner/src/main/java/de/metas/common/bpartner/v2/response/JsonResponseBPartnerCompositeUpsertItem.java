/*
 * #%L
 * de-metas-common-bpartner
 * %%
 * Copyright (C) 2021 metas GmbH
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program. If not, see
 * <http://www.gnu.org/licenses/gpl-2.0.html>.
 * #L%
 */

package de.metas.common.bpartner.v2.response;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonInclude.Include;
import com.fasterxml.jackson.annotation.JsonProperty;
import io.swagger.v3.oas.annotations.media.Schema;
import lombok.Builder;
import lombok.Singular;
import lombok.Value;

import javax.annotation.Nullable;
import java.util.List;

@Value
public class JsonResponseBPartnerCompositeUpsertItem
{
	@Schema
	@JsonInclude(Include.NON_NULL)
	JsonResponseUpsertItem responseBPartnerItem;

	@Schema
	@JsonInclude(Include.NON_EMPTY)
	List<JsonResponseUpsertItem> responseLocationItems;

	@Schema
	@JsonInclude(Include.NON_EMPTY)
	List<JsonResponseUpsertItem> responseContactItems;

	@Schema
	@JsonInclude(Include.NON_EMPTY)
	List<JsonResponseUpsertItem> responseBankAccountItems;

	@Schema
	@JsonInclude(Include.NON_EMPTY)
	List<JsonResponseUpsertItem> responseCreditLimitItems;

	@Builder
	@JsonCreator
	public JsonResponseBPartnerCompositeUpsertItem(
			@JsonProperty("responseBPartnerItem") @Nullable final JsonResponseUpsertItem responseBPartnerItem,
			@JsonProperty("responseLocationItems") @Singular final List<JsonResponseUpsertItem> responseLocationItems,
			@JsonProperty("responseContactItems") @Singular final List<JsonResponseUpsertItem> responseContactItems,
			@JsonProperty("responseBankAccountItems") @Singular final List<JsonResponseUpsertItem> responseBankAccountItems,
			@JsonProperty("responseCreditLimitItems") @Singular final List<JsonResponseUpsertItem> responseCreditLimitItems)
	{
		this.responseBPartnerItem = responseBPartnerItem;
		this.responseLocationItems = responseLocationItems;
		this.responseContactItems = responseContactItems;
		this.responseBankAccountItems = responseBankAccountItems;
		this.responseCreditLimitItems = responseCreditLimitItems;
	}

}
