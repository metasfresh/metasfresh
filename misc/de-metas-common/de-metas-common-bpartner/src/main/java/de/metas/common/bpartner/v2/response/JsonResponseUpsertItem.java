/*
 * #%L
 * de-metas-common-bpartner
 * %%
 * Copyright (C) 2021 metas GmbH
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program. If not, see
 * <http://www.gnu.org/licenses/gpl-2.0.html>.
 * #L%
 */

package de.metas.common.bpartner.v2.response;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import de.metas.common.rest_api.common.JsonMetasfreshId;
import de.pentabyte.springfox.ApiEnum;
import io.swagger.annotations.ApiModel;
import io.swagger.annotations.ApiModelProperty;
import lombok.Builder;
import lombok.NonNull;
import lombok.Value;

import javax.annotation.Nullable;

@ApiModel("Response to a a single mater data upsert request entity")
@Value
public class JsonResponseUpsertItem
{
	public enum SyncOutcome
	{
		CREATED,

		@ApiEnum("Master data was updated; note that it's possible that nothing really changed due to the update.")
		UPDATED,

		@ApiEnum("E.g. if a location already exists and the sync advise is to not change existing entities.")
		NOTHING_DONE
	}

	@ApiModelProperty(value = "The identifier that was specified in the repective upsert request",//
			position = 10)
	String identifier;

	@ApiModelProperty(value = "The metasfresh-ID of the upserted record.\n"
			+ "Can be null if the respective resource did not exist and the sync-advise indicated to do nothing.",//
			position = 20, dataType = "java.lang.Long")
	JsonMetasfreshId metasfreshId;

	SyncOutcome syncOutcome;

	@Builder
	@JsonCreator
	public JsonResponseUpsertItem(
			@JsonProperty("identifier") @NonNull final String identifier,
			@JsonProperty("metasfreshId") @Nullable final JsonMetasfreshId metasfreshId,
			@JsonProperty("syncOutcome") @NonNull final SyncOutcome syncOutcome)
	{
		this.identifier = identifier;
		this.metasfreshId = metasfreshId;
		this.syncOutcome = syncOutcome;
	}

}
