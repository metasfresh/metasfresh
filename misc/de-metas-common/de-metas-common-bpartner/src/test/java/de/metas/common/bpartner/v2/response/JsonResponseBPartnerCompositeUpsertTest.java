/*
 * #%L
 * de-metas-common-bpartner
 * %%
 * Copyright (C) 2021 metas GmbH
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program. If not, see
 * <http://www.gnu.org/licenses/gpl-2.0.html>.
 * #L%
 */

package de.metas.common.bpartner.v2.response;

import com.fasterxml.jackson.databind.ObjectMapper;
import de.metas.common.bpartner.v1.response.JsonResponseBPartnerCompositeUpsert;
import de.metas.common.bpartner.v1.response.JsonResponseBPartnerCompositeUpsertItem;
import de.metas.common.bpartner.v1.response.JsonResponseUpsertItem;
import de.metas.common.rest_api.common.JsonMetasfreshId;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import java.io.IOException;

import static io.github.jsonSnapshot.SnapshotMatcher.expect;
import static io.github.jsonSnapshot.SnapshotMatcher.start;
import static org.assertj.core.api.Assertions.assertThat;

class JsonResponseBPartnerCompositeUpsertTest
{
	ObjectMapper objectMapper = new ObjectMapper();
	
	@BeforeAll
	public static void beforeAll()
	{
		start();
	}

	@Test
	public void serializeDeserializeTest() throws IOException
	{
		final JsonResponseBPartnerCompositeUpsert item = JsonResponseBPartnerCompositeUpsert.builder()
				.responseItem(JsonResponseBPartnerCompositeUpsertItem.builder()
						.responseBPartnerItem(JsonResponseUpsertItem.builder()
								.identifier("val-12345")
								.metasfreshId(JsonMetasfreshId.of(23))
								.syncOutcome(JsonResponseUpsertItem.SyncOutcome.CREATED)
								.build())
						.responseContactItem(JsonResponseUpsertItem.builder()
								.identifier("ext-12345")
								.metasfreshId(JsonMetasfreshId.of(24))
								.syncOutcome(JsonResponseUpsertItem.SyncOutcome.UPDATED)
								.build())
						.responseLocationItem(JsonResponseUpsertItem.builder()
								.identifier("ext-12345")
								.metasfreshId(JsonMetasfreshId.of(24))
								.syncOutcome(JsonResponseUpsertItem.SyncOutcome.UPDATED)
								.build())
						.responseBankAccountItem(JsonResponseUpsertItem.builder()
								.identifier("ext-12345")
								.metasfreshId(JsonMetasfreshId.of(24))
								.syncOutcome(JsonResponseUpsertItem.SyncOutcome.NOTHING_DONE)
								.build())
						.build())
				.build();
		
		final String str = objectMapper.writeValueAsString(item);

		final JsonResponseBPartnerCompositeUpsert result = objectMapper.readValue(str,JsonResponseBPartnerCompositeUpsert.class);
		assertThat(result).isEqualTo(item);

		expect(result).toMatchSnapshot();
	}

}
