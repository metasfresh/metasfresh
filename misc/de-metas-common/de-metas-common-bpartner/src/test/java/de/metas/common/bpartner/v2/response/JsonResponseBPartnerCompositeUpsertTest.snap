de.metas.common.bpartner.v2.response.JsonResponseBPartnerCompositeUpsertTest.serializeDeserializeTest=[
  {
    "responseItems": [
      {
        "responseBPartnerItem": {
          "identifier": "val-12345",
          "metasfreshId": 23,
          "syncOutcome": "CREATED"
        },
        "responseBankAccountItems": [
          {
            "identifier": "ext-12345",
            "metasfreshId": 24,
            "syncOutcome": "NOTHING_DONE"
          }
        ],
        "responseContactItems": [
          {
            "identifier": "ext-12345",
            "metasfreshId": 24,
            "syncOutcome": "UPDATED"
          }
        ],
        "responseLocationItems": [
          {
            "identifier": "ext-12345",
            "metasfreshId": 24,
            "syncOutcome": "UPDATED"
          }
        ]
      }
    ]
  }
]